starky2/src/proof.rs

```diff
@@ -0,0 +1,259 @@
+    pub permutation_zs_cap: Option<MerkleCapTarget>,
+    pub quotient_polys_cap: MerkleCapTarget,
+    pub openings: StarkOpeningSetTarget<D>,
+    pub opening_proof: FriProofTarget<D>,
+}
+
+impl<const D: usize> StarkProofTarget<D> {
+    /// Recover the length of the trace from a STARK proof and a STARK config.
+    pub fn recover_degree_bits(&self, config: &StarkConfig) -> usize {
+        let initial_merkle_proof = &self.opening_proof.query_round_proofs[0]
+            .initial_trees_proof
+            .evals_proofs[0]
+            .1;
+        let lde_bits = config.fri_config.cap_height + initial_merkle_proof.siblings.len();
+        lde_bits - config.fri_config.rate_bits
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct StarkProofWithPublicInputs<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    pub proof: StarkProof<F, C, D>,
+    // TODO: Maybe make it generic over a `S: Stark` and replace with `[F; S::PUBLIC_INPUTS]`.
+    pub public_inputs: Vec<F>,
+}
+
+pub struct StarkProofWithPublicInputsTarget<const D: usize> {
+    pub proof: StarkProofTarget<D>,
+    pub public_inputs: Vec<Target>,
+}
+
+pub struct CompressedStarkProof<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    /// Merkle cap of LDEs of trace values.
+    pub trace_cap: MerkleCap<F, C::Hasher>,
+    /// Purported values of each polynomial at the challenge point.
+    pub openings: StarkOpeningSet<F, D>,
+    /// A batch FRI argument for all openings.
+    pub opening_proof: CompressedFriProof<F, C::Hasher, D>,
+}
+
+pub struct CompressedStarkProofWithPublicInputs<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    pub proof: CompressedStarkProof<F, C, D>,
+    pub public_inputs: Vec<F>,
+}
+
+pub(crate) struct StarkProofChallenges<F: RichField + Extendable<D>, const D: usize> {
+    /// Randomness used in any permutation arguments.
+    pub permutation_challenge_sets: Option<Vec<GrandProductChallengeSet<F>>>,
+
+    /// Random values used to combine STARK constraints.
+    pub stark_alphas: Vec<F>,
+
+    /// Point at which the STARK polynomials are opened.
+    pub stark_zeta: F::Extension,
+
+    pub fri_challenges: FriChallenges<F, D>,
+}
+
+pub(crate) struct StarkProofChallengesTarget<const D: usize> {
+    pub permutation_challenge_sets: Option<Vec<GrandProductChallengeSet<Target>>>,
+    pub stark_alphas: Vec<Target>,
+    pub stark_zeta: ExtensionTarget<D>,
+    pub fri_challenges: FriChallengesTarget<D>,
+}
+
+/// Purported values of each polynomial at the challenge point.
+#[derive(Debug, Clone)]
+pub struct StarkOpeningSet<F: RichField + Extendable<D>, const D: usize> {
+    pub local_values: Vec<F::Extension>,
+    pub next_values: Vec<F::Extension>,
+    pub permutation_ctl_zs: Option<Vec<F::Extension>>,
+    pub permutation_ctl_zs_right: Option<Vec<F::Extension>>,
+    pub ctl_zs_last: Vec<F>,
```

Comment:
This is the same poly as the two fields above (at a different point) right? If so I think the names should be similar.

Reply:
So this is the openings of the cross-table lookups `Z` polynomials at the last point of the trace, whereas the openings above are for the permutation + cross-table lookups `Z` polynomials.
That's why I only kept the `ctl_zs` part of the name `permutation_ctl_zs`.

Reply:
Oh makes sense now, I just didn't know to read it as permutation *and* ctl. The names seem fine then, I would just comment to make the "and" and the meaning of ctl clear.

Reply:
Sounds good, I added some comments 👍